Source/Events/EventConverter.cs

```
@@ -73,28 +96,36 @@ public Contracts.UncommittedAggregateEvents ToProtobuf(UncommittedAggregateEvent
-                    source.ExternalEventReceived.ToDateTimeOffset(),
-                    @event);
-            }
-            else
-            {
-                return new CommittedEvent(
-                    source.EventLogSequenceNumber,
-                    source.Occurred.ToDateTimeOffset(),
-                    source.EventSourceId.To<EventSourceId>(),
-                    source.ExecutionContext.ToExecutionContext(),
-                    @event);
-            }
+                var artifact = ToSDK(source.Type);
+                try
+                {
+                    var @event = _serializer.JsonToEvent(artifact, source.Content);
+                    if (source.External)
+                    {
+                        return new CommittedExternalEvent(
+                            source.EventLogSequenceNumber,
+                            source.Occurred.ToDateTimeOffset(),
+                            source.EventSourceId.To<EventSourceId>(),
+                            source.ExecutionContext.ToExecutionContext(),
+                            source.ExternalEventLogSequenceNumber,
+                            source.ExternalEventReceived.ToDateTimeOffset(),
+                            @event);
+                    }
+                    else
+                    {
+                        return new CommittedEvent(
+                            source.EventLogSequenceNumber,
+                            source.Occurred.ToDateTimeOffset(),
+                            source.EventSourceId.To<EventSourceId>(),
+                            source.ExecutionContext.ToExecutionContext(),
+                            @event);
+                    }
+                }
+                catch (UnableToInstantiateInstanceOfType ex)
+                {
+                    _logger.Warning(ex, "Couldn't deserialize artifact: {Artifact} to a committed event.", artifact);
+                    throw;
+                }
```

There are a couple of things here I'm thinking of. 
* The variable `artifact`, it's not really artifact, it is the `eventType`. I know it's not your code but would be nice to change it here :) 
* We could have a little bit more developer friendly log message by using this format instead `Couldn't deserialize artifact '{Artifact}' to '{EventType}' event` where `Artifact` is `source.Type.Id.To<ArtifactId>`
* We should throw an exception here, but I don't think that we should throw the original one. Maybe we should have a `CouldNotDeserializeEvent` that takes an `ArtifactId` and the `Type` of the `Event`? And also I think we can drop giving the exception `ex` as an argument to `_logger.Warning`

---

@jakhog  opinions? As you have more experience actually using this. What's your take?

---

First of all, I would change the caught exception type to just `Exception` since more things can go wrong here (JSON syntax errors e.g.), and in stead of logging create the exception that you mention @woksin, but also send in the actual JSON content of the event, and its position in the event store (scope and sequence number).

---

That’s a good point with the either throw or log. We should have some guidelines around this down the line.

Also agree on the JSON content part